### PR TITLE
[webdriver]: click elements in Shadow DOM

### PR DESCRIPTION
**Pull request by Linkgoron**
I've added tests for clicking elements inside of shadow DOM in the WebDriver tests.

Currently they don't pass in WebKit ([Bug](https://bugs.webkit.org/show_bug.cgi?id=209233)), but pass in Firefox and Chrome.

I'd be happy to get some pointers to maybe make the tests a bit less verbose.

### CODE REVIEW

File: webdriver/tests/element_click/shadow_dom.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+def test_inside_element_click(session):
+    session.url = get_checkbox_dom()
+    custom_element = session.find.css("custom-checkbox-element", all=False)
+    checkbox_element = session.execute_script("return arguments[0].shadowRoot.querySelector('input')", args=(custom_element,))
+    is_pre_checked = session.execute_script("return arguments[0].checked", args=(checkbox_element,))
+    assert is_pre_checked == False
+    click_response = element_click(session, checkbox_element)
+    assert_success(click_response)
+    is_post_checked = session.execute_script("return arguments[0].checked", args=(checkbox_element,))
+    assert is_post_checked == True
```
Comment (gsnedders) [inline webdriver/tests/element_click/shadow_dom.py:50]: You could do something with [parametrize](https://docs.pytest.org/en/latest/parametrize.html), like `@pytest.mark.parametrize("click_on", ["custom_element", "checkbox_element"])` and then I guess `element_click(session, locals()[click_on])`, which eliminates basically all the repetition. Not sure there's anything much better than that, given we do need to refer to all the elements.

**Review by gsnedders — COMMENTED** [review]
Thanks!

This basically looks fine, see the one comment for a suggestion on how to improve it based on your question.

Couple of other nits: you're inconsistent as to whether there's one or two new lines between functions (mostly we follow PEP8 and go for two), and I'd put all the args to the "execute_script" calls on their own line on help with reading them, like:

```python
session.execute_script(
    "return arguments[0].shadowRoot.querySelector('input').checked",
    args=(custom_element,)
)
```

File: webdriver/tests/element_click/shadow_dom.py (hunk 1 above)
Comment (Linkgoron) [reply webdriver/tests/element_click/shadow_dom.py]: Thanks! This cut it down to two tests with parameters.

**Review by gsnedders — APPROVED** [review]
Thanks! This looks great now.

That said, I'm not quite the most competent reviewer around this part of WPT, so I'll leave it open till Monday evening and land it then if nobody else objects.

**Comment by whimboo** [comment]
Well, we don't have support for Shadow DOM in the WebDriver specification yet. So how can we land tests that trigger a click on nodes within a Shadow DOM? 

Also leaving the PR open over the weekend will not give many people a chance to reply. 

@AutomatedTester, @jgraham is that PR ok to leave in or shall we back it out?

**Comment by jgraham** [comment]
If the shadow dom is open it seems logical that this ought to work. Closed shadow dom would be harder since, as you say we don't have commands for piercing the shadow dom boundary. But I think this is OK.

**Comment by gsnedders** [comment]
> Well, we don't have support for Shadow DOM in the WebDriver specification yet. So how can we land tests that trigger a click on nodes within a Shadow DOM?

It is clearly what users expect. It is what Chrome and Firefox already implement, and it is considered a bug in Safari. Are you seriously debating whether the behaviours this expects are what the behaviour should be?

> Also leaving the PR open over the weekend will not give many people a chance to reply.

It was open for most of Monday, and I did solicit any further opinions on IRC before landing this.

**Comment by burg** [comment]
Several bugs about this issue have been filed on safaridriver, so its behavior is being adjusted to match Chrome/Firefox. It would be great for the spec to hint that this is supposed to work. I've waited a few years for clarity and have given up.

**Pull request merged**